Add CI infrastructure, fix some nits

slotThe opened the pull request:
Stack builds: https://github.com/slotThe/X11-xft/actions/runs/1458776382
Haskell CI builds: https://github.com/slotThe/X11-xft/actions/runs/1458776381

In order to get https://github.com/xmonad/xmonad-contrib/pull/646 going more quickly, I decided to fudge around with the CI for this repo.  I also used this opportunity to clean things up a little.

Obligatory reminder that I have no idea what I'm doing

## Commits

##### [Sync Cabal version with Hackage](4b396d1f6e045e5379dfc5e0c59c620875909511)

The newest version on Hackage is 0.3.1, while this is being declared as
0.3.3 in the cabal file.  This might be confusing to... everyone
involved and so it seems best to sync these.

##### [Add CHANGES.md](1d268d280087af98884a81aa0252d7a4c5c224dd)

##### [Add README.md](1467db80191d76620fff7ee75db2f5337f7c819c)

##### [Update cabal file](b759fec6c0c7283ebb8d64a1397706fc4e1aa58d)

* Remove spurious `Hello' executable, since this is strictly a library
  and as such this is unused.
* Change the source-repository URL to the xmonad project—the repo got
  transferred to us by Clemens.
* Delete the unused Setup.lhs.

##### [X11-xft.cabal: Add -fpedantic, update default flags](f41dff55cf76afc63e5efd573588e7bc9467d2cd)

* Add -Wall and -fno-warn-unused-binds to the default ghc-options.  This
  is to keep this in sync with the X11 library
* Much like the rest of the project, introduce a `pedantic' flag that
  enables -Werror (and, for now, nothing else).  This will later be used
  in the CI to ensure that new pull requests do not introduce any new
  warnings.

##### [Add stack.yaml](4f7252eb77a19c339effc4b0436146926c9e1bb1)

Add a simple stack.yaml file to make the project easily compilable via
stack.

##### [Silence warnings](0be3422bffc83f1b656c341dccb01e68be26fe49)

With the new default build flags introduced in [1], we now get a lot of
warnings that need to be fixed before enabling the CI; do that.

[1]: 39fa9335638a6086ba00fbf3bed98e1e69fadce0
     (X11-xft.cabal: Add -fpedantic, update default flags)

##### [Add GitHub actions](74a9b20d9b58db353a8e406a1ab7b707f15bdb85)

This is mostly a copy-paste of what X11 does[1].

[1]: https://github.com/xmonad/X11/tree/master/.github/workflows

##### [Update gitignore](82b6c9f7933d03906b882584f7b76416117ae9ef)

Ignore the usual suspects when it comes to building with Cabal and
stack.

slotThe committed:
Add CHANGES.md

slotThe committed:
Add README.md

slotThe committed:
Update cabal file

* Remove spurious `Hello' executable, since this is strictly a library
  and as such this is unused.
* Change the source-repository URL and the current maintainer to the
  xmonad project—the repo got transferred to us by Clemens.
* Delete the unused Setup.lhs.

slotThe committed:
X11-xft.cabal: Add -fpedantic, update default flags

* Add -Wall and -fno-warn-unused-binds to the default ghc-options.  This
  is to keep this in sync with the X11 library
* Much like the rest of the project, introduce a `pedantic' flag that
  enables -Werror (and, for now, nothing else).  This will later be used
  in the CI to ensure that new pull requests do not introduce any new
  warnings.

slotThe committed:
Add stack.yaml

Add a simple stack.yaml file to make the project easily compilable via
stack.

slotThe committed:
Silence warnings

With the new default build flags introduced in [1], we now get a lot of
warnings that need to be fixed before enabling the CI; do that.

[1]: 39fa9335638a6086ba00fbf3bed98e1e69fadce0
     (X11-xft.cabal: Add -fpedantic, update default flags)

slotThe committed:
Add GitHub actions

This is mostly a copy-paste of what X11 does[1].

[1]: https://github.com/xmonad/X11/tree/master/.github/workflows

Co-authored-by: Tomas Janousek <tomi@nomi.cz>

slotThe committed:
Update gitignore

Ignore the usual suspects when it comes to building with Cabal and
stack.

slotThe commented:
The fact that we have an empty `configure.ac` is obviously quite horrible (and just a stupid hack to make haskell-ci happy) and should probably be changed in some way.  I, however, have absolutely no idea what that file even is (nor do I understand what the X11 version does)

liskin commented on X11-xft.cabal:
```diff
@@ -1,6 +1,6 @@
 cabal-version: 2.2
 name:          X11-xft
-version:       0.3.3
+version:       0.3.1
```
You'd need to revert https://github.com/xmonad/X11-xft/commit/0f0066a71eef5ac8d879c84a42f0622a38cc12fa for this to make sense

liskin commented on CHANGES.md:
```diff
@@ -0,0 +1,7 @@
+
+## unknown
+
+## 0.3.1 (2021-05-09)
```
0.3.1 was released in 2011

slotThe replied:
Oh, indeed, I was thinking that 0f0066a already made it to Hackage as 0.3.1.  Should've read more carefully.  In that case the version number should stay at 0.3.3 as well, I suppose

geekosaur commented:
`configure.ac` is needed in the `X11` package to find the various X11 extensions or complain if they're not present. If `X11-xft` doesn't need to do this (which it probably doesn't) then we should adapt the ci-actions to not use `configure.ac`. (I could look, but I have no idea how github actions works.)

clefru commented:
Thanks for taking care of this package. Before we accept contributions, it might make sense to switch the license to BSD3. The library does not create enough value on its own to give future users a license headache. I don't feel strongly about it, but in case somebody else does: "I as the original author support relicensing to BSD3."

liskin commented:
> Thanks for taking care of this package. Before we accept contributions, it might make sense to switch the license to BSD3. The library does not create enough value on its own to give future users a license headache. I don't feel strongly about it, but in case somebody else does: "I as the original author support relicensing to BSD3."

Indeed, if you still hold all the copyrights to allow relicensing, it's probably best to do it. Very good point, thanks!

liskin reviewed (approved):
I pushed a fix for the haskell-ci and packdeps workflows and added README/CHANGES to the sdist tarball, the rest looks okay. Thanks.
Consider changing the license to BSD3 as @clefru said, then we can rebase and merge.

clefru commented:
> Indeed, if you still hold all the copyrights to allow relicensing, it's probably best to do it. Very good point, thanks!

I just review my old darcs repo, and I do hold all the copyright.

slotThe committed:
Relicense to BSD3

This is to stay in line with all of the other xmonad repositories.  This
was okayed by Clemens, who holds all of the copyright.

  > Before we accept contributions, it might make sense to switch the
    license to BSD3. […] "I as the original author support relicensing
    to BSD3." [1]

  > I just review my old darcs repo, and I do hold all the
    copyright. [2]

[1]: https://github.com/xmonad/X11-xft/pull/1#issuecomment-968294839
[2]: https://github.com/xmonad/X11-xft/pull/1#issuecomment-968370355

Co-authored-by: Clemens Fruhwirth <clemens@endorphin.org>
Signed-off-by: Clemens Fruhwirth <clemens@endorphin.org>

slotThe commented:
It hurts me a little bit to relicense something from copyleft to permissive, but if that's what the majority wants I won't object.

@clefru are you happy with https://github.com/xmonad/X11-xft/pull/1/commits/1032ad31b61f8ed85a95e09f4c4c0e2da41101f2 ?

clefru commented:
> It hurts me a little bit to relicense something from copyleft to permissive, but if that's what the majority wants I won't object.

I do strongly prefer copyleft. For cryptsetup, I long refused GPL->LGPL relicensing, because there is real value in that software. But how much value you does X11-xft bring to the table? X11-xft is a simple wrapper. There are maybe 20 LOC that meet the bar of copyrightable work. If the LGPL license is an obstacle, somebody can rewrite it in half day, and that somebody will be grumpy about it. In this case, I prefer to fit in with the X11 and Xft license. 

> @clefru are you happy with [1032ad3](https://github.com/xmonad/X11-xft/commit/1032ad31b61f8ed85a95e09f4c4c0e2da41101f2) ?

Yes, I am. Thank you!
